http_test.go

```diff
@@ -26,3 +79,139 @@ func Test_urlToFilename(t *testing.T) {
+				CopyFile(cache.GetFileName("sample.pdf"), "fixtures/sample.pdf", 0644)
+
+				req := httptest.NewRequest("GET", "/documents/somewhere/sample.pdf?page=10", nil)
+				recorder := httptest.NewRecorder()
+
+				h.handleImage(recorder, req)
+				So(recorder.Result().StatusCode, ShouldEqual, 404)
+			})
+
+			Convey("When the page is not valid", func() {
+				os.MkdirAll(filepath.Join(os.TempDir(), filepath.Dir(filename)), 0755)
+				CopyFile(cache.GetFileName("sample.pdf"), "fixtures/sample.pdf", 0644)
+
+				req := httptest.NewRequest("GET", "/documents/somewhere/sample.pdf?page=-1", nil)
+				recorder := httptest.NewRecorder()
+
+				h.handleImage(recorder, req)
+				So(recorder.Result().StatusCode, ShouldEqual, 400)
+			})
+
+			Convey("When file is not present", func() {
+				req := httptest.NewRequest("GET", "/documents/somewhere/asdf.pdf", nil)
+				recorder := httptest.NewRecorder()
+				h.handleImage(recorder, req)
+
+				body, err := ioutil.ReadAll(recorder.Result().Body)
+				So(err, ShouldBeNil)
+				So(recorder.Result().StatusCode, ShouldEqual, 400)
+				So(string(body), ShouldContainSubstring, "Invalid page")
+			})
+
+			Convey("Rejects badly/unsigned URLs when signing is required", func() {
+				req := httptest.NewRequest("GET", "/documents/somewhere/sample.pdf?page=1", nil)
+				recorder := httptest.NewRecorder()
+				h.urlSecret = "secret"
+
+				h.handleImage(recorder, req)
+
+				body, err := ioutil.ReadAll(recorder.Result().Body)
+				So(err, ShouldBeNil)
+				So(recorder.Result().StatusCode, ShouldEqual, 403)
+				So(string(body), ShouldContainSubstring, "Invalid signature")
+			})
+		})
+
+		Convey("When everything is working", func() {
+				os.MkdirAll(filepath.Join(os.TempDir(), filepath.Dir(filename)), 0755)
+				CopyFile(cache.GetFileName("sample.pdf"), "fixtures/sample.pdf", 0644)
+				recorder := httptest.NewRecorder()
+
+			Convey("Handles a normal request", func() {
+				req := httptest.NewRequest("GET", "/documents/somewhere/sample.pdf?page=1", nil)
+
+				h.handleImage(recorder, req)
+
+				body, err := ioutil.ReadAll(recorder.Result().Body)
+				So(err, ShouldBeNil)
+				So(len(body), ShouldBeGreaterThan, 1024)
+				So(recorder.Result().StatusCode, ShouldEqual, 200)
+				So(didDownload, ShouldBeTrue)
+				So(downloadCount, ShouldEqual, 1)
+			})
+
+			Convey("Handles a jpegs", func() {
+				req := httptest.NewRequest("GET", "/documents/somewhere/sample.pdf?page=1&width=1024&quality=75&imageType=image/jpeg", nil)
+
+				h.handleImage(recorder, req)
+
+				body, err := ioutil.ReadAll(recorder.Result().Body)
+				So(err, ShouldBeNil)
+				So(len(body), ShouldBeGreaterThan, 1024)
+				So(recorder.Result().StatusCode, ShouldEqual, 200)
+				So(recorder.Result().Header["Content-Type"][0], ShouldEqual, "image/jpeg")
+			})
+
+			Convey("Handles a pngs", func() {
+				req := httptest.NewRequest("GET", "/documents/somewhere/sample.pdf?page=1&width=1024&quality=75&imageType=image/png", nil)
```

Comment:
Somehow, I am surprised that the slash in `image/png` doesn't need to be URL encoded.